File: 13.transparent-origami/hs/part2.hs

import System.IO
import Data.List.Split
import Data.List
import Data.Set (Set)
import qualified Data.Set as Set

applyFold points (d, v) =
    if d == 'x' then
        map (\(x, y) -> if x > v then (2*v - x, y) else (x, y)) points
    else
        map (\(x, y) -> if y > v then (x, 2*v - y) else (x, y)) points

tupleDo f (a, b) (c, d) = (f a c, f b d)

main = do
    dat <- readFile "input.txt"
    let (rawpoints:rawfolds:_) = splitOn [""] (splitOn "\n" dat)
    let points = map ((\(x:y:_) -> (x, y)) . map (read :: String -> Int) . splitOn ",") rawpoints
    let folds = map ((\(d:v:_) -> (last d, read v :: Int)) . splitOn "=") rawfolds
    let folded = Set.fromList (foldl applyFold points folds)
    let minp = foldl (tupleDo min) (Set.elemAt 1 folded) folded
    let maxp = foldl (tupleDo max) (Set.elemAt 1 folded) folded
    putStrLn ("Result: \n" ++ 
        intercalate "\n" [
            intercalate "" [
                if Set.member (x, y) folded then "##" else "  "
                | x <- [fst minp..fst maxp]
            ] | y <- [snd minp..snd maxp]
        ])

File: 13.transparent-origami/py/part2.py

import sys

from functools import reduce

raw = sys.stdin.read()

rawpoints, rawfolds = raw.split('\n\n')

points = set((int(p.split(',')[0]), int(p.split(',')[1])) for p in rawpoints.split('\n'))
folds = [ (f.split('=')[0][-1], int(f.split('=')[1])) for f in rawfolds.strip().split('\n') ]

def applyFold(points, fold):
    if fold[0] == 'x':
        return set((x, y) if x < fold[1] else (2 * fold[1] - x, y) for x, y in points)
    elif fold[0] == 'y':
        return set((x, y) if y < fold[1] else (x, 2 * fold[1] - y) for x, y in points)

for fold in folds:
    points = applyFold(points, fold)

min = reduce(lambda a, x: (min(a[0], x[0]), min(a[1], x[1])), points)
max = reduce(lambda a, x: (max(a[0], x[0]), max(a[1], x[1])), points)

print('Result:')

for y in range(min[1], max[1] + 1):
    for x in range(min[0], max[0] + 1):
        print('██' if (x, y) in points else '  ', end='')
    print()

